internal/fs/inode/type_cache.go

```diff
@@ -50,58 +87,126 @@ type typeCache struct {
 	//
 	// INVARIANT: entries.CheckInvariants() does not panic
 	// INVARIANT: Each value is of type cacheEntry
-	entries lrucache.Cache
+	entries *lru.Cache
 }
 
-// Create a cache whose information expires with the supplied TTL. If the TTL
-// is zero, nothing will ever be cached.
-func newTypeCache(perTypeCapacity int, ttl time.Duration) typeCache {
-	return typeCache{
-		ttl:     ttl,
-		entries: lrucache.New(perTypeCapacity),
+// NewTypeCache creates an LRU-policy-based cache with given max-size and TTL.
+// Any entry whose TTL has expired, is removed from the cache on next access (Get).
+// When insertion of next entry would cause size of cache > sizeInMB,
+// older entries are evicted according to the LRU-policy.
+// If any of TTL or sizeInMB is zero, nothing is ever cached
+//
+//	i.e. Insert, Erase do nothing, and Get will always return UnknownType.
+func NewTypeCache(sizeInMB int, ttl time.Duration) TypeCache {
+	if ttl > 0 && sizeInMB != 0 {
+		if sizeInMB < -1 {
+			panic("unhandled scenario: type-cache-max-size-mb < -1")
+		}
+		var lruSizeInBytesToUse uint64 = math.MaxUint64 // default for when sizeInMb = -1
+		if sizeInMB > 0 {
+			lruSizeInBytesToUse = util.MiBsToBytes(uint64(sizeInMB))
+		}
+		return &typeCache{
+			ttl:     ttl,
+			entries: lru.NewCache(lruSizeInBytesToUse),
+		}
 	}
-}
-
-////////////////////////////////////////////////////////////////////////
-// Public interface
-////////////////////////////////////////////////////////////////////////
-
-// Panic if any internal invariants have been violated. The careful user can
-// arrange to call this at crucial moments.
-func (tc *typeCache) CheckInvariants() {
-	tc.entries.CheckInvariants()
+	return &typeCache{}
 }
 
 // Insert inserts a record to the cache.
 func (tc *typeCache) Insert(now time.Time, name string, it Type) {
-	// Are we disabled?
-	if tc.ttl == 0 {
-		return
+	if tc.entries != nil {
+		_, err := tc.entries.Insert(name, cacheEntry{
+			expiry:    now.Add(tc.ttl),
+			inodeType: it,
+		})
+		if err != nil {
+			panic(fmt.Errorf("failed to insert entry in typeCache: %v", err))
+		}
+		logger.Debugf("TypeCache: Inserted %s as %s", name, it.String())
```

Comment:
this will increase the size of debug logs a lot and we haven't planned functional testing so far for metadata cache, so please remove this and other debug statements in type cache.